metagamer/environments/tictactoe.py

```diff
@@ -0,0 +1,176 @@
+        for turn in turn_order:
+            yield turn
+
+
+def check_win(board: np.array) -> int:
+    """
+    Given a 3*3 numpy array whose elements are 0, 1, -1
+    representing empty, player1 and player 2,
+    return the player who's won if any or 0 if no winner found
+
+    Args:
+        board (np.array): 3 x 3 board
+
+    Returns:
+        int: 0  -> no winner
+             1  -> player 1
+             -1 -> player 2
+
+    """
+    assert board.shape == (3, 3)
+    for axis in [0, 1]:
+        axis_sum = board.sum(axis=axis)
+        if 3 in axis_sum:
+            return 1
+        elif -3 in axis_sum:
+            return -1
+
+    diag = board.diagonal().sum()
+    diag_inv = np.fliplr(board).diagonal().sum()
+    for diag_sum in [diag, diag_inv]:
+        if diag_sum == 3:
+            return 1
+        elif diag_sum == -3:
+            return -1
+
+    # no winner yet
+    return 0
+
+
+def to_one_hot(board: np.array) -> np.array:
+    """
+    Convert the representation to one hot encoding
+    We assume a 3x3 board input
+
+    Args:
+        board (np.array): 3 x 3
+
+    Returns:
+        3 x 3 x 3:
+            board[:, :  0] = location of empty squares
+            board[:, :, 1] = location of player 1
+            board[:, :, 2] = location of player 2 moves
+
+    """
+    oh = np.stack((board == 0, board == 1, board == -1))
+    return oh.shape.astype(int)
+
+
+def to_human(board: np.array) -> np.array:
+    """Convert this into a """
+    human_board = np.full(board.shape, " ")
+    human_board[np.where(board == 1)] = "X"
+    human_board[np.where(board == -1)] = "O"
+    return human_board
+
+
+class TicTacToeEnv(gym.Env):
+    """
+    TicTacToe environment in the openai gym style
+
+    This doesn't support observation_space but we can adapt that later
+
+    """
+
+    metadata = {"render.modes": ["human"]}
+
+    TURN_ORDER = [1, -1]
+
+    def __init__(self):
+        self.board_shape = 3, 3
+        self.symbols = {1: "X", -1: "O"}
+        self.action_space = spaces.Tuple([spaces.Discrete(3), spaces.Discrete(3)])
+        # we don't support this right now
+        self.observation_space = None
+
+        self.board = None
+        self.turn_iterator = None
+        self.curr_turn = None
+        self.done = None
+
+        self.reset()
+
+    def reset(self) -> np.array:
+        self.board = np.zeros(self.board_shape, dtype=int)
+        self.turn_iterator = turn_iterator(self.TURN_ORDER)
+        self.curr_turn = next(self.turn_iterator)
+        self.done = False
+        return self._get_obs()
+
+    def _get_obs(self) -> np.array:
+        """
+        Abstracted the observation from the underlying state though in this case they're
+        identical. This is useful if doing something like changing the underlying state
+        or converting to one hot encoding
+
+        Returns:
+            np.array of 3x3 representing the board in it's default state
+
+        """
+        return self.board
```

Comment:
Not quite sure why this function exists, I assume it's a requirement for the class as deterined by the gym style?

Reply:
I think the docstring was meant to explain it. But the point is that we return the observation from reset and step. This means if we want to change how we return state, we just change this function as opposed to editing step and reset.

This effectively means doing different observation spaces are as easy as sublcasses while not doing so would require wrappers.  It's a common design pattern you see in gym environments and it's effectively a bit of future-proofing by giving you more flexibility.

Reply:
Ah ok.. Sounds like a good idea. So would we could have different versions of `_get_obs` in future?  Would this be good if we wanted to make more complicated derived versions observed state, relative to true state? (like some of the board is always hidden, for example)?

Reply:
Pretty much - you can do this via gym wrappers, but I think a subclass is easier. I saw this a lot in third party gym environments and it makes a lot of sense.

I've updated the doc again, but this might just be a trust me for now